clippy_lints/src/dbg_macro.rs

```diff
@@ -31,85 +32,95 @@ declare_clippy_lint! {
     "`dbg!` macro is intended as a debugging tool"
 }
 
-#[derive(Copy, Clone)]
+#[derive(Clone)]
 pub struct DbgMacro {
     allow_dbg_in_tests: bool,
+    /// Keep tracks the `dbg!` macro callsites that are already checked,
+    /// so that we can save some performance by skipping any expressions from the same expansion.
+    checked_dbg_call_site: FxHashSet<Span>,
 }
 
 impl_lint_pass!(DbgMacro => [DBG_MACRO]);
 
 impl DbgMacro {
     pub fn new(allow_dbg_in_tests: bool) -> Self {
-        DbgMacro { allow_dbg_in_tests }
+        DbgMacro {
+            allow_dbg_in_tests,
+            checked_dbg_call_site: FxHashSet::default(),
+        }
     }
 }
 
 impl LateLintPass<'_> for DbgMacro {
     fn check_expr(&mut self, cx: &LateContext<'_>, expr: &Expr<'_>) {
-        let Some(macro_call) = root_macro_call_first_node(cx, expr) else {
+        let Some(macro_call) =
+            root_macro_call(expr.span).filter(|mc| cx.tcx.is_diagnostic_item(sym::dbg_macro, mc.def_id))
```

Comment:
This should find the first `dbg!` macro in the chain, not the root macro. Expansions of `macro_rules! foo { () => (dbg!()) }` still need to lint.